src/android/java/io/jxcore/node/BtConnectorHelper.java

```diff
@@ -232,18 +241,21 @@ public void BeginConnectPeer(String toPeerId, ConnectStatusCallback connectStatu
         }
     }
 
+    //this is always called in context of thread that created instance of the library
     @Override
     public void Connected(BluetoothSocket bluetoothSocket, boolean incoming,String peerId,String peerName,String peerAddress) {
 
         if(bluetoothSocket != null) {
 
             AddPeerIfNotDiscovered(bluetoothSocket, peerId, peerName, peerAddress);
-            print_debug("Starting the connected thread incoming : " + incoming);
+            print_debug("Starting the connected thread incoming : " + incoming + ", " +peerName);
 
             if (incoming) {
                 BtToServerSocket tmpBtToServerSocket = new BtToServerSocket(bluetoothSocket, this);
-                mServerSocketList.add(tmpBtToServerSocket);
 
+                synchronized(mServerSocketList) {
+                    mServerSocketList.add(tmpBtToServerSocket);
+                }
```

Comment:
@DrJukka Line 271 below is not thread safe because between the null check on 271 and the stop on 272 someone else could have changed mBtToRequestSocket to be null. Please first copy to a local variable and then check for null and call stop.